### ISSUE TITLE
CQRS infrastructure

### ISSUE THREAD

**Issue by hikalkan** [abpframework/abp#57]
Investigate the EventFlow framework:

Source: https://github.com/eventflow/EventFlow
Document: http://docs.geteventflow.net/

**Comment by tomlewandowski** [comment]
What do you think about this?
https://www.youtube.com/watch?v=_lwCVE_XgqI
https://github.com/JasonGT/NorthwindTraders

and using CQRS + MediatR?

**Comment by hikalkan** [comment]
Thanks @tomlewandowski for sharing. Will take a look.

**Comment by kegumx** [comment]
HI, Is there any progress？

**Comment by hikalkan** [comment]
Not worked on this yet.

**Comment by GFoley83** [comment]
Command / Query segregation via MediatR is the only thing I've _really_ been waiting on for ABP.

Some refactoring of the event bus and background job service layers could also be done to leverage the [MediatR notification pipeline](https://github.com/jbogard/MediatR/wiki#publishing) for event dispatching.

**Comment by tomlewandowski** [comment]
@hikalkan 
I found a very nice repository that is a very good reference to DDD and CQRS.
https://github.com/kgrzybek/modular-monolith-with-ddd
What do you think about this two building blocks?:
https://github.com/kgrzybek/modular-monolith-with-ddd/blob/master/src/BuildingBlocks/Domain/TypedIdValueBase.cs

https://github.com/kgrzybek/modular-monolith-with-ddd/blob/master/src/BuildingBlocks/Domain/IBusinessRule.cs

This is another great resource:
https://herbertograca.com/2017/11/16/explicit-architecture-01-ddd-hexagonal-onion-clean-cqrs-how-i-put-it-all-together/

**Comment by pranavpatil19** [comment]
set property high and please  move to milestone 1.0

**Comment by hikalkan** [comment]
Can't move it to 1.0, it is very close to release 1.0 and we can't add new features to 1.0.

**Comment by pranavpatil19** [comment]
is this module move to 1.1

**Comment by hikalkan** [comment]
We will see, can't promise.

**Comment by pranavpatil19** [comment]
. i request you because i want to integrate CQRS With Rabbitmq .  if you take this high priority after 1.0 release its really helpful

**Comment by GFoley83** [comment]
@tomlewandowski that video you linked to was absolute gold and I think perfectly showcased how well MediatR would slot into the current ABP solution design. It's a perfect fit.

**Comment by ramax495** [comment]
Any progress on CQRS + MediatR ?

**Comment by pranavpatil19** [comment]
i am also waiting for this module....  2year old request.

**Comment by hikalkan** [comment]
>i am also waiting for this module.... 2year old request.

Yes, but v1.0 has been released just 1 month ago :) We don't want to work on new features while leaving current feature incomplete & undocumented. Built-in CQRS infrastructure on the road map, but can't promise on any date.

**Comment by thiag0coelho** [comment]
Is there room for this feature in 2.2?

**Comment by maskalek** [comment]
Hey guys. Any progress/news? Considering option to migrate on your framework and but we would really love to keep working with MediatR.

**Comment by maulik-modi** [comment]
Any update on MediatR?

**Comment by kegumx** [comment]
+1

**Comment by jader1313** [comment]
I too would appreciate the implementation of CQRS + MediatR

**Comment by feynmanloo** [comment]
Any progress on CQRS + MediatR ?

**Comment by tunechr** [comment]
Hi Guys, Just looking at an evaluating abp.io I've used zero in the past and really liked it

Can I check where this feature thread is in planning? 

For my clarification, as CQRS and Event Sourcing are related but not the same... but we are discussing both here (and clean architecture) so I'm a bit confused as to what this feature would be?

I'd love to get to the point where we have an event sourcing model, or at least a stronger event model which can then feed into features such as **automation** and **workflows**, as well as opening up options for CQRS for developers.

I believe this is a missing enabler for the abp core platform and it's something that I've done manually at the moment using MediatR and a lot of bespoke code in abp.zero... so hoping that can be part of the new platform :)

Imagine being able to use something like elsa workflow / camunda (or your own tool) to be able to allow users to create their own actions and triggers and create automation such as :

- When [event] then [action]
- When someone creates a user then add them to the onboarding workflow
- When someone creates a tenant then send an email to the marketing team
- When someone marks an invoice paid then send a message in teams to accounting. 
or
To be able to build out a BPMN workflow using these events and activities. 

Just my thoughts :)

**Comment by barnybadzoo** [comment]
CQRS + MediatR.. any plan to implement this soon? Is there a roadmap so see when its planned ? I am also evaluating abp currently.. Would be a great step ahead if you plan this

**Comment by maulik-modi** [comment]
@barnybadzoo, @tomlewandowski  and @hikalkan  , have you checked https://github.com/ardalis/ApiEndpoints shows MediatR is optional

**Comment by barnybadzoo** [comment]
@maulik-modi , yes i did check the Link..but ApiEndpoints with MediatR reduces complexity compared to the well known MVC pattern. Of course you can have both. But please check this FAQ: https://github.com/dotnet-architecture/eShopOnWeb/wiki/Frequently-Asked-Questions where Microsoft shows for both an example (OrderController and ManageController). From my point of view ApiEndpoint with MediatR is much simpler and groups everything that has to do with a particular API endpoint together.

**Comment by maulik-modi** [comment]
@barnybadzoo , there are 3 choices - 
1) Use controller - NO MediatR - ManageController
2) Use controller + MediatR - OrderController
3) Use ApiEndpoints - NO controller, NO MediatR - https://github.com/ardalis/ApiEndpoints

I am referring to 3)

**Comment by xNarkon** [comment]
Any update? What about intermodular communication using the CQRS pattern? Currently, there is no possibility to get some resources from another module in the monolith architecture but there are clear needs sometimes to fetch some data. In the microservices architecture, this is typically normal and there are possibilities to do that, using REST API / Query Bus, etc. But using modules and building modular monolith using ready-to-use components there is no possibility to easily do that without introducing coupling

**Comment by yangfan-zz** [comment]
Simple implementation of abp cqrs, using MediatR [Fan.Abp.Cqrs](https://github.com/yangfan-zz/Fan.Abp)

**Comment by appsupport-gc** [comment]
I found it fairly simple to use the nuget package FS.Abp.MediatR.Domain and use normal MediatR Logic.


In SomeDomainModule.cs:
 public override void ConfigureServices(ServiceConfigurationContext context)
    {
        context.Services.AddAbpMediatR&lt;SomeDomainModule&gt;();
  }


 public class CreateSomethingCommand : IRequest&lt;Something&gt;
{
   .....
{

 public class CreateSomethingCommandHandler : IRequestHandler&lt;CreateSomethingCommand, Something&gt;
    {
async Task&lt;Something&gt; IRequestHandler&lt;CreateSomethingCommand, Something&gt;.Handle(CreateSomethingCommand request, CancellationToken cancellationToken)
        {
             return new Something();
        }
}

To call inject IMediator mediator and then call 
mediator.Send(new CreateSomethingCommand([arguments])

It would be nice if there was an officially supported Volo.Abp.MediatR.Domain nuget package

**Comment by adam-dot-cohen** [comment]
Any updates on this?

**Comment by johnnypea** [comment]
I would be interested in seeing this in action as well. Any pointers?

**Comment by sponge19990501** [comment]
Hi, any updates on this issue?

**Comment by hikalkan** [comment]
No, this is not planned yet. To be honest, we don't think that we can invest our time on that in a short time.

Actually, we would like to hear from the community about their experiences on implementing CQRS in their ABP based solutions. We want to learn what is the problem of using existing CQRS libraries in your ABP solutions. Please write to us about your experiences.

**Comment by sponge19990501** [comment]
Understood. In reality, I haven't used any CQRS package or methods on ABP before.

The reason ABP team wants to implement this is to have an officially designed solution to reduce trial and error.

However, it sounds like this might not happen quickly, so I'll try using CQRS on ABP and provide feedback or contributions.

**Comment by hikalkan** [comment]
To be honest, our team has no expertise on CQRS implementation in real world scenarios. We are mostly in theorical side. So, it would be good to know experience from community (comments, articles, draft integrations, etc) before we deep dive that topic. After all, this is a community driven project :)

**Comment by aelhadi** [comment]
Some disadvantages of Event Sourcing with CQRS are:

It can be complex to implement and maintain
It can require more disk space and memory than traditional data stores
It can have longer bootup times after failures or upgrades.
It can lead to eventual consistency issues in some applications.
It can be challenging to handle entities with long and complex lifespans

I'm not fun of CQRS, it's a waste of time in my opinion ;-)  a lot of complexity for nothing...

**Comment by PiyushGulhane-05** [comment]
Please check this out, this could be of great help as well: https://www.udemy.com/course/net-microservices-cqrs-event-sourcing-with-kafka/

**Comment by PGW5** [comment]
> To be honest, our team has no expertise on CQRS implementation in real world scenarios. We are mostly in theorical side. So, it would be good to know experience from community (comments, articles, draft integrations, etc) before we deep dive that topic. After all, this is a community driven project :)

Please check this out, this could be of great help (great architecture implementation), https://www.udemy.com/course/net-microservices-cqrs-event-sourcing-with-kafka/

**Comment by aelhadi** [comment]
> > To be honest, our team has no expertise on CQRS implementation in real world scenarios. We are mostly in theorical side. So, it would be good to know experience from community (comments, articles, draft integrations, etc) before we deep dive that topic. After all, this is a community driven project :)
> 
> Please check this out, this could be of great help (great architecture implementation), https://www.udemy.com/course/net-microservices-cqrs-event-sourcing-with-kafka/

it's a waste of time in my opinion ;-)

**Comment by adam-dot-cohen** [comment]
CQRS adds complexity as aptly put by @aelhadi.  But the complexity of CQRS offers substantial ROI that increases as a function of the scale - and the inherent complexity that comes with it.

If we consider all of the points made by @aelhadi, they apply to all distributed system architectures  - especially microservices which add an array items to @aelhadi's CQRS list above.  My personal favorite is the art of partitioning the problem domain (reminds me of Zen and the Art of Motorcycle Maintenance) - and the data integrity and eventual consistency problems that follow.

ABP's microservice implementation is a good example.  Breaking your system into pieces across domain and network boundaries equates to more things to maintain, more resources to utilize, more things to fail, more data to synchronize - and the path to precision analytics and business intelligence can be a long road...in some cases to nowhere.

In my opinion, any attempt to implement CQRS on top ABP framework's existing architecture is a fools errand.  On the other hand, an in process mediator with decoupled decorator support such as MediatR - would add substantial value in the form of collapsing the project system (9+ projects is a bit much), decoupling the kitchen sink of dependencies, and providing extensibility w/o the headache of modules (and the 9 additional projects per module).

**Comment by MarkLFT** [comment]
I have used cqrs and mediatr on a number of api projects and found it extremely helpful in isolating code and removing dependencies. What I find most useful is using mediatr for inter-module requests in a modular monolith design. It is extremely lightweight and as it is in-process, very fast. Whilst completely removing the need for any inter-module dependencies.

I often check back to the abp documentation to see if the mediatr pattern is supported yet, as for me this is a big roadblock.  Which is a shame because I like so many other things this platform includes.
